Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm

```diff
@@ -0,0 +1,348 @@
+    else {
+        auto sessionRange = contextRangeForTextIndicatorStyle(uuid);
+
+        if (!sessionRange)
+            sessionRange = contextRangeForTextIndicatorStyle(uuid);
+
+        if (!sessionRange) {
+            completionHandler();
+            return;
+        }
+        document->markers().addTransparentContentMarker(*sessionRange, uuid);
+    }
+    completionHandler();
+}
+
+
+void TextIndicatorStyleController::createTextIndicatorForRange(const WebCore::SimpleRange& range, CompletionHandler<void(std::optional<WebCore::TextIndicatorData>&&)>&& completionHandler)
+{
+    if (!m_webPage) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    RefPtr corePage = m_webPage->corePage();
+    if (!corePage) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    if (RefPtr localMainFrame = dynamicDowncast<WebCore::LocalFrame>(corePage->mainFrame())) {
+        std::optional<WebCore::TextIndicatorData> textIndicatorData;
+        constexpr OptionSet textIndicatorOptions {
+            WebCore::TextIndicatorOption::IncludeSnapshotOfAllVisibleContentWithoutSelection,
+            WebCore::TextIndicatorOption::ExpandClipBeyondVisibleRect,
+            WebCore::TextIndicatorOption::UseSelectionRectForSizing,
+            WebCore::TextIndicatorOption::SkipReplacedContent,
+            WebCore::TextIndicatorOption::RespectTextColor
+        };
+        if (auto textIndicator = WebCore::TextIndicator::createWithRange(range, textIndicatorOptions, WebCore::TextIndicatorPresentationTransition::None, { }))
+            textIndicatorData = textIndicator->data();
+        completionHandler(WTFMove(textIndicatorData));
+        return;
+    }
+    completionHandler(std::nullopt);
+}
+
+void TextIndicatorStyleController::createTextIndicatorForID(const WTF::UUID& uuid, CompletionHandler<void(std::optional<WebCore::TextIndicatorData>&&)>&& completionHandler)
+{
+    auto sessionRange = contextRangeForTextIndicatorStyle(uuid);
+
+    if (!sessionRange) {
+        completionHandler(std::nullopt);
+        return;
+    }
+    createTextIndicatorForRange(*sessionRange, WTFMove(completionHandler));
+}
+
+void TextIndicatorStyleController::enableTextIndicatorStyleAfterElementWithID(const String& elementID, const WTF::UUID& uuid)
+{
+    RefPtr corePage = m_webPage->corePage();
+    if (!corePage) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+    RefPtr frame = corePage->checkedFocusController()->focusedOrMainFrame();
+    if (!frame) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    RefPtr document = frame->document();
+    if (!document) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    RefPtr root = document->documentElement();
+    if (!root) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    WebCore::VisibleSelection fullDocumentSelection(WebCore::VisibleSelection::selectionFromContentsOfNode(root.get()));
+    auto simpleRange = fullDocumentSelection.range();
```

Comment:
should this be `firstRange`

Reply:
I don't know what you mean by this.

Reply:
instead of
```
auto simpleRange = fullDocumentSelection.range();
```


should it be
```
auto simpleRange = fullDocumentSelection.range();
```

? (which is what we do in UnifiedTextReplacementController

Reply:
Those... are the same line... what did you mean for the second one?

Reply:
lol oops, I meant

```
auto simpleRange = fullDocumentSelection.firstRange();
```